src/Microsoft.Azure.WebJobs.Host/Extensions/Tooling.cs

```diff
@@ -0,0 +1,397 @@
+                }
+            }
+        }
+
+        private bool CanBind(Attribute attribute, Type parameterType)
+        {
+            IBindingProvider root = this._root;
+
+            bool result = Task.Run(() => ScriptHelpers.CanBindAsync(root, attribute, parameterType)).GetAwaiter().GetResult();
+
+            return result;
+        }
+               
+        // Get a better implementation 
+        public Type GetDefaultType(
+            Attribute attribute,
+            FileAccess access, // direction In, Out, In/Out
+            Type requestedType) // combination of Cardinality and DataType            
+        {
+            if (attribute == null)
+            {
+                throw new ArgumentNullException(nameof(attribute));
+            }
+            if (access == FileAccess.Write)
+            {
+                var outType = _outTypes.FirstOrDefault(type => this.CanBind(attribute, type));
+                if (outType != null)
+                {
+                    return outType;
+                }
+                // Error! 
+                throw new InvalidOperationException($"Can't bind {attribute.GetType().Name} to a script-compatible  output.");
+            }
+            else if (access == FileAccess.Read)
+            {
+                if (IsBinary(requestedType))
+                {
+                    return typeof(byte[]);
+                }
+
+                var array = IsCardinalityMany(requestedType);
+                if (array != null)
+                {
+                    var inner = this.GetDefaultType(attribute, FileAccess.Read, array);
+                    if (inner != null)
+                    {
+                        return inner.MakeArrayType();
+                    }
+                }
+
+                var inType = _inTypes.FirstOrDefault(type => this.CanBind(attribute, type));
+                if (inType != null)
+                {
+                    return inType;
+                }
+                throw new InvalidOperationException($"Can't bind {attribute.GetType().Name} to a script-compatible input.");
+            }
+            else
+            {
+                throw new NotImplementedException($"Can't bind {attribute.GetType().Name} to an In/Out parameter.");
+            }
+        }
+
+        // Return null if cardinality is singular. 
+        // Return the element type if cardinality is Many
+        private static Type IsCardinalityMany(Type requestedType)
+        {
+            if (IsBinary(requestedType))
+            {
+                return null;
+            }
+            if (requestedType.IsArray)
+            {
+                return requestedType.GetElementType();
+            }
+            return null;
+        }
+
+        private static bool IsBinary(Type requestedType)
+        {
+            return requestedType == typeof(byte[]);
+        }
+
+        // Helpers for testing some Script Scenarios 
+        public static class ScriptHelpers
```

Comment:
I'll change to private. This is only used as a helper in this class. I prefer a private nested class since that limits the bread of access.

---
In reply to: [108534601](https://github.com/Azure/azure-webjobs-sdk/pull/1093#discussion_r108534601) [](ancestors = 108534601)